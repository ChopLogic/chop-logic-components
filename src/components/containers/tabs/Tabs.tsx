import React, { useState } from 'react';

import ChopLogicTabContent from './elements/TabContent';
import ChopLogicTabList from './elements/TabList';
import { StyledTabsContainer } from './Tabs.styled';
import { ChopLogicTabsProps } from './types';

const ChopLogicTabs: React.FC<ChopLogicTabsProps> = ({ tabs, defaultTabId, mode = 'horizontal', ...rest }) => {
  const tabIds = tabs.map((item) => item.id);
  const tabPanelIds = tabIds.map((id) => `tabpanel_${id}`);
  const defaultId = defaultTabId && tabIds.includes(defaultTabId) ? defaultTabId : tabIds[0];
  const [selectedTabId, setSelectedTabId] = useState(defaultId);

  const handleTabSelect = (id: string) => {
    setSelectedTabId(id);
  };

  return (
    <StyledTabsContainer $mode={mode} {...rest}>
      <ChopLogicTabList
        tabs={tabs}
        selectedTabId={selectedTabId}
        tabPanelIds={tabPanelIds}
        onTabSelect={handleTabSelect}
        mode={mode}
        tabIds={tabIds}
      />
      <ChopLogicTabContent tabs={tabs} selectedTabId={selectedTabId} />
    </StyledTabsContainer>
  );
};

export default ChopLogicTabs;
