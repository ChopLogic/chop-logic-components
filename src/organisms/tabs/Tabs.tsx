import { OrientationMode } from '@enums';
import { TabsProps } from '@models';
import { getClassName } from '@utils';
import { FC, useState } from 'react';

import { TabContent } from './content/TabContent';
import { TabList } from './list/TabList';
import styles from './Tabs.module.scss';

const Tabs: FC<TabsProps> = ({
  tabs,
  defaultTabId,
  mode = OrientationMode.Horizontal,
  className,
  stretched,
  editable = false,
  onTabTitleChange,
  extendable = false,
  onTabAdd,
  extendedTabLabel = 'New Tab',
  extendedTabContent,
  ...rest
}) => {
  const tabIds = tabs.map((item) => item.id);
  const tabPanelIds = tabIds.map((id) => `tabpanel_${id}`);
  const defaultId = defaultTabId && tabIds.includes(defaultTabId) ? defaultTabId : tabIds[0];
  const [selectedTabId, setSelectedTabId] = useState(defaultId);
  const tabsClass = getClassName([styles.tabs, className, { [styles.tabs__vertical]: mode === OrientationMode.Vertical }]);

  const handleTabSelect = (id: string) => {
    setSelectedTabId(id);
  };

  const handleTabTitleChange = (tabId: string, newTitle: string) => {
    onTabTitleChange?.(tabId, newTitle);
  };

  const handleTabAdd = () => {
    onTabAdd?.();
  };

  return (
    <div {...rest} className={tabsClass}>
      <TabList
        tabs={tabs}
        selectedTabId={selectedTabId}
        tabPanelIds={tabPanelIds}
        onTabSelect={handleTabSelect}
        onTabTitleChange={handleTabTitleChange}
        mode={mode}
        tabIds={tabIds}
        stretched={stretched}
        editable={editable}
        extendable={extendable}
        onTabAdd={handleTabAdd}
        extendedTabLabel={extendedTabLabel}
      />
      <TabContent tabs={tabs} selectedTabId={selectedTabId} extendedTabContent={extendedTabContent} extendable={extendable} />
    </div>
  );
};

export default Tabs;
