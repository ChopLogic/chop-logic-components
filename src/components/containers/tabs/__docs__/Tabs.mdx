import { Meta } from '@storybook/blocks';
import * as ChopLogicTabsStories from './Tabs.stories';

<Meta of={ChopLogicTabsStories} title='Tabs' />

# ChopLogicTabs

Versatile tab component that allows you to switch between different pieces of content by clicking on corresponding tabs.

### Usage

```tsx
import React from 'react';
import ChopLogicTabs from 'chop-logic-components';

const tabs = [
  { id: 'tab1', title: 'Tab 1', content: <div>Content for Tab 1</div> },
  { id: 'tab2', title: 'Tab 2', content: <div>Content for Tab 2</div> },
  { id: 'tab3', title: 'Tab 3', content: <div>Content for Tab 3</div> },
];

const ExampleComponent = () => <ChopLogicTabs tabs={tabs} defaultTabId='tab1' />;

export default ExampleComponent;
```

### Props

| Prop             | Type                  | Description                                                                                                      | Default        | Required |
|------------------|-----------------------|------------------------------------------------------------------------------------------------------------------|----------------|----------|
| **tabs**         | `ChopLogicTabItem[]`  | An array of tab items, each containing an id, title, and the content to display.                                 | `[]`           | Yes      |
| **defaultTabId** | `string`              | The id of the tab that should be active by default.                                                              | `undefined`    | No       |
| **mode**         | `ChopLogicTabsMode`   | The orientation of the tabs.                                                                                     | `'horizontal'` | No       |
| **id**           | `string`              | The id global attribute defines an identifier (ID) which must be unique in the whole document.                   | `undefined`    | No       |
| **className**    | `string`              | The class global attribute is a list of the classes of the element.                                              | `undefined`    | No       |
| **style**        | `CSSProperties`       | The style global attribute contains CSS styling declarations to be applied to the element.                       | `undefined`    | No       |
| **tabIndex**     | `number`              | The tabindex global attribute allows developers to make HTML elements focusable.                                 | `undefined`    | No       |
| **title**        | `string`              | The title global attribute contains text representing advisory information related to the element it belongs to. | `undefined`    | No       |
