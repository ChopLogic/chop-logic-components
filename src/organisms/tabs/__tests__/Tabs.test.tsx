import { OrientationMode } from '@enums';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it } from 'vitest';

import Tabs from '../Tabs';

describe('Tabs', () => {
  const tabs = [
    { id: 'tab-1', title: 'Tab 1', disabled: false, content: <>Tab 1 content</> },
    { id: 'tab-2', title: 'Tab 2', disabled: false, content: <>Tab 2 content</> },
    { id: 'tab-3', title: 'Tab 3', disabled: false, content: <>Tab 3 content</> },
  ];

  const testProps = {
    tabs,
    defaultTabId: 'tab-1',
    mode: OrientationMode.Horizontal,
    className: 'test-class',
    id: 'tabs-id',
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<Tabs {...testProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render another tab content when user clicks on tabs', async () => {
    render(<Tabs {...testProps} />);
    expect(screen.getByText('Tab 1 content')).toBeInTheDocument();

    await userEvent.click(screen.getByText('Tab 2'));
    expect(screen.queryByText('Tab 1 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 2 content')).toBeInTheDocument();

    await userEvent.click(screen.getByText('Tab 3'));
    expect(screen.queryByText('Tab 2 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 3 content')).toBeInTheDocument();

    await userEvent.click(screen.getByText('Tab 1'));
    expect(screen.queryByText('Tab 3 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 1 content')).toBeInTheDocument();
  });

  it('should render another tab content when user switches tabs via keyboard', async () => {
    render(<Tabs {...testProps} />);
    expect(screen.getByText('Tab 1 content')).toBeInTheDocument();

    const firstTab = screen.getByText('Tab 1');
    firstTab.focus();
    await userEvent.keyboard('[ArrowRight]');
    expect(screen.queryByText('Tab 1 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 2 content')).toBeInTheDocument();

    await userEvent.keyboard('[ArrowRight]');
    expect(screen.queryByText('Tab 2 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 3 content')).toBeInTheDocument();

    await userEvent.keyboard('[ArrowRight]');
    expect(screen.queryByText('Tab 3 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 1 content')).toBeInTheDocument();

    await userEvent.keyboard('[ArrowLeft]');
    expect(screen.queryByText('Tab 1 content')).not.toBeInTheDocument();
    expect(screen.getByText('Tab 3 content')).toBeInTheDocument();
  });
});
