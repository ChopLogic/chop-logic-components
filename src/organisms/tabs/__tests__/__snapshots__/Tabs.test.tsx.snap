// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="_tabs_19db3f test-class"
    id="tabs-id"
  >
    <div
      class="_tabList_7b8ac2"
      data-testid="tab-list"
      role="tablist"
      tabindex="0"
    >
      <span
        class="_tab_wrapper_54926d _tab_wrapper__selected_54926d _tab_wrapper__stretched_54926d"
      >
        <button
          aria-controls="tabpanel_tab-1"
          aria-selected="true"
          class="_tab_54926d"
          id="tab-1"
          role="tab"
        >
          Tab 1
        </button>
      </span>
      <span
        class="_tab_wrapper_54926d _tab_wrapper__stretched_54926d _tab_wrapper__inactive_54926d"
      >
        <button
          aria-controls="tabpanel_tab-2"
          aria-selected="false"
          class="_tab_54926d"
          id="tab-2"
          role="tab"
        >
          Tab 2
        </button>
      </span>
      <span
        class="_tab_wrapper_54926d _tab_wrapper__stretched_54926d _tab_wrapper__inactive_54926d"
      >
        <button
          aria-controls="tabpanel_tab-3"
          aria-selected="false"
          class="_tab_54926d"
          id="tab-3"
          role="tab"
        >
          Tab 3
        </button>
      </span>
    </div>
    <div
      aria-labelledby="tab-1"
      class="_tabContent_7c3dbf"
      data-testid="tab-content"
      id="tabpanel_tab-1"
      role="tabpanel"
    >
      Content 1
    </div>
  </div>
</DocumentFragment>
`;
